napari/_app_model/utils.py

```diff
@@ -0,0 +1,92 @@
+from typing import Union
+
+from app_model.expressions import parse_expression
+from app_model.types import Action, MenuItem, SubmenuItem
+
+from napari._app_model import get_app
+from napari._app_model.constants import MenuGroup, MenuId
+
+MenuOrSubmenu = Union[MenuItem, SubmenuItem]
+
+
+def contains_dummy_action(menu_items: list[MenuOrSubmenu]) -> bool:
+    """True if one of the menu_items is the dummy action, otherwise False.
+
+    Parameters
+    ----------
+    menu_items : list[MenuOrSubmenu]
+        menu items belonging to a given menu
+
+    Returns
+    -------
+    bool
+        True if menu_items contains dummy item otherwise false
+    """
+    for item in menu_items:
+        if hasattr(item, 'command') and 'empty_dummy' in item.command.id:
+            return True
+    return False
+
+
+def is_empty_menu(menu_id: str) -> bool:
+    """Return True if the given menu_id is empty, otherwise False
+
+    Parameters
+    ----------
+    menu_id : str
+        id of the menu to check
+
+    Returns
+    -------
+    bool
+        True if the given menu_id is empty, otherwise False
+    """
+    app = get_app()
+    if menu_id not in app.menus:
+        return True
+    if len(app.menus.get_menu(menu_id)) == 0:
+        return True
+    if len(app.menus.get_menu(menu_id)) == 1 and contains_dummy_action(
+        app.menus.get_menu(menu_id)
+    ):
+        return True
+    return False
+
+
+def no_op() -> None:
+    """Fully qualified no-op to use for dummy actions."""
+
+
+def get_dummy_action(id_key: str, menu_id: MenuId) -> tuple[Action, str]:
+    """Return a dummy action to be used for the given menu.
+
+    id_key will be formatted into the action ID: 'napari.{id_key}.empty_dummy'.
+
+    Parameters
+    ----------
+    id_key: str
+        key to be formatted into the action ID and when expression
+    menu_id: MenuId
+        id of the menu to add the dummy action to
+
+    Returns
+    -------
+    tuple[Action, str]
+        dummy action and the `when` expression context key
+    """
+    action = Action(
+        id=f'napari.{id_key}.empty_dummy',
+        title='Empty',
+        callback=no_op,
+        menus=[
+            {
+                'id': menu_id,
+                'group': MenuGroup.NAVIGATION,
+                # parse_expression can't take a variable name, so we
+                # walrus context_key here to be able to return it
+                'when': parse_expression(context_key := f'{id_key}_empty'),
+            }
+        ],
+        enablement=False,
+    )
+    return action, context_key
```

Comment:
I super-applaud the use of "walrus" as a verb 👏, but I'm not 100% clear on what you are saying. Are you saying that this wouldn't work???

```suggestion
    context_key = f'{id_key}_empty'
    action = Action(
        id=f'napari.{id_key}.empty_dummy',
        title='Empty',
        callback=no_op,
        menus=[
            {
                'id': menu_id,
                'group': MenuGroup.NAVIGATION,
                # parse_expression can't take a variable name, so we
                # walrus context_key here to be able to return it
                'when': parse_expression(context_key),
            }
        ],
        enablement=False,
    )
    return action, context_key
```
???

Because that is pretty wild and definitely challenges my understanding of how Python and the walrus work!

Also note: I am not saying reformat the code, I'm saying, I'm not 100% sure what the comment means so maybe it needs additional explaining, maybe a link to some docs, etc.

Reply:
> Are you saying that this wouldn't work???

yep, that is exactly right. The expression needs to be either an `Expression` object, or a string

Reply:
... But (pure) Python expressions are [referentially transparent](https://stackoverflow.com/questions/210835/what-is-referential-transparency), so that shouldn't matter, and indeed it seems like it doesn't?

```
In [1]: from app_model.expressions import parse_expression
In [2]: expr_str = 'a > 2'
In [3]: parse_expression(expr_str)
Out[3]: Expr.parse('a > 2')

In [4]: parse_expression('a > 2')
Out[4]: Expr.parse('a > 2')
```

Reply:
(Just to chat: I don't know this particular case, but there are cases where this is not true: whenever meta-programming/code inspection are involved :P For example, how new layers come up with a name based on the input variable name.)

Reply:
@jni I see what you're saying, though I'm not familiar with referential transparency, but when I tried it the way you suggest (which I did before the walrus), napari fully froze after I added an item to the previously empty menu and tried to look at `Layers` menu again.

And I thought the expr parsing might be the problem, so swapped it out and it worked.

Reply:
> (Just to chat: I don't know this particular case, but there are cases where this is not true: whenever meta-programming/code inspection are involved :P For example, how new layers come up with a name based on the input variable name.)

Yeah, that was the "pure" 😅 but it doesn't quite encapsulate the restrictions. I previously had (most) in the parentheses and I should have stuck with that. 😂

And anyway, that's why I actually went and tested it in IPython in case something funky was happening with `parse_expression`.

... which is sounds like it might be, though I'm more inclined to think that some *other* (stochastic) funny business was going on @DragaDoncila. But, given my code above, I think I'll (sadly) remove the walrus comment, as it doesn't seem to capture what is actually going wrong...

> I'm not familiar with referential transparency

it just means that:

```
name = some * complicated + expression('foo')
do_thing_with(name)
```

should be equivalent to

```
do_thing_with(some * complicated + expression('foo'))
```

Giving something a name (reference) or not should not affect what the code does. (ie should be transparent).

Reply:
> Giving something a name (reference) or not should not affect what the code does. (ie should be transparent).

Ah ok yeah, I think that's a fantastic concept in general, but I'm not surprised it's breaking down with the level of magic and indirection we have going on here...

Reply:
This is the `parse_expression` [function](https://github.com/pyapp-kit/app-model/blob/main/src/app_model/expressions/_expressions.py#L40) btw - I don't really get it to be truthful, but it might elucidate things for you:


```python
def parse_expression(expr: Expr | str) -> Expr:
    """Parse string expression into an [`Expr`][app_model.expressions.Expr] instance.

    Parameters
    ----------
    expr : Expr | str
        Expression to parse.  (If already an `Expr`, it is returned)

    Returns
    -------
    Expr
        Instance of `Expr`.

    Raises
    ------
    SyntaxError
        If the provided string is not an expression (e.g. it's a statement), or
        if it uses any forbidden syntax components (e.g. Call, Attribute,
        Containers, Indexing, Slicing, f-strings, named expression,
        comprehensions.)
    """
    if isinstance(expr, Expr):
        return expr
    try:
        # mode='eval' means the expr must consist of a single expression
        tree = ast.parse(str(expr), mode="eval")
        if not isinstance(tree, ast.Expression):
            raise SyntaxError  # pragma: no cover
        return ExprTransformer().visit(tree.body)
    except SyntaxError as e:
        raise SyntaxError(f"{expr!r} is not a valid expression: ({e}).") from None
```

Reply:
Yeah none of that is doing anything that would break when you pass in a variable pointing to a string rather than a literal string.

Hmm I'm worried that the hang you saw is some weird stochastic thing that happens regardless of whether you walrus it up. Actually I had to restart a job to get it green, and now I see that the failed job failed after hanging:

```
  napari\_qt\widgets\_tests\test_qt_dims.py ..........s               [3660/4195] [0:06:21]
  napari\_qt\widgets\_tests\test_qt_dims_2.py ..
  Error: The action 'Test with tox main' has timed out after 40 minutes.
```

([here](https://github.com/napari/napari/actions/runs/9815321407/job/27104420024?pr=7038)).

How do you test adding things to the menu? (ie how should I manually run this a few times locally to make sure that this isn't a recurrent problem?)

Reply:
Ruh-roh! @jni check the PR description for a script u can pop in the console to register an action with a menu. The `register_action` call returns a function u can call to deregister the action, so if u save that and call it u can do a full round trip

Reply:
> The register_action call returns a function u can call to deregister the action, so if u save that and call it u can do a full round trip

Thanks! This was super useful so I updated the example in the description. But anyway, I tried over and over in different menus and it's working great, so imma merge! Thank you @DragaDoncila!